### ISSUE TITLE
Trace compaction doesn't apply to inner layers

### ISSUE THREAD

**Issue by frankmcsherry** [TimelyDataflow/differential-dataflow#205]
Trace compaction currently occurs only for the largest layer in the trace, an artifact of driving the process by volume of updates: if we compacted inner layers we may fail to initiate merges for larger layers if the working set never grows in size.

This has negative implications for memory and for performance. As a simple example, consider a program that maintains `distinct()` for a collection of numbers whose input is fed by:
```rust
for round in 0 .. {
    input.advance_to(round);
    input.insert(round % large);
    input.insert(round % small);
}
```
Take `large` to be 1M and `small` to be 1: the largest layer will eventually have 1M entries in it, and some 500K updates to the small set will still be making their way through small layers. Importantly, they will be uncompacted, so any attempts to *use* them will have a cost proportional to the ratio between `large` and `small`, rather than logarithmic in the number of updates.

This cost gets paid loading the data from the trace, at which point it is compacted, so this cost is felt most by small batch, low latency settings.

If we hold `large` at 1M and vary the batch size from 1,000 down, swinging through 10M updates for different `small` takes time:

| batch | `small`: 1 | `small`: 1,000 |
| ------: | -------: |----:|
| 1,000 | 81.48s | 50.84s |
| 100 | 467.07s | 56.39s |
| 10 | 3406.77s | 106.22s |

Clearly smaller batches for the same number of updates could take more time, but ideally it is clear that there is a gap between that effect when observed with `small` at 1,000 and with `small` at 1.

**Comment by frankmcsherry** [comment]
Preliminary work is being done in #204 which enables compaction of all layers without impeding merging of later layers. Early results on the same benchmark (`examples/compact.rs`) are:

| batch | `small`: 1 | `small`: 1,000 |
|---:|---:|---:|
|1,000 | 18.23s | 26.12s |
| 100 | 33.73s | 38.48s |
| 10 | 126.26s | 129.30s |

These results are a bit fouled by in-progress nature of the PR. Ideally we can see that it is less sensitive to `small` though, and if we can keep that and engineer the rough corners down we should be a bit happier.

Interestingly, we are faster for large batch sizes which isn't what I expected at all.

**Comment by frankmcsherry** [comment]
This is addressed in #204 which has been merged. It is not yet the default behavior, but we'll start trying it out now.
